source/blog/2017-12-13-compose-custom-ostree.md

```diff
@@ -0,0 +1,185 @@
+
+Install Docker and run its service on the system where you are planning to run the container:
+
+```
+$ sudo dnf install docker
+$ sudo systemctl start docker
+```
+
+## Dockerfile
+Create a Dockerfile with the following content which we will use to build a Docker container using Fedora27 as its base and on top of it some additional setup is being done.
+
+```
+$ cat Dockerfile
+FROM fedora:27
+
+# Update, install specified packages and clean cached information
+RUN dnf update -y && dnf install -y git python rpm-ostree; \
+dnf clean all
+
+# Create specified directories
+RUN mkdir -p /srv /srv/repo /srv/cache
+
+# Set /srv as working directory and clone fedora-atomic repository into it
+WORKDIR /srv
+RUN git clone https://pagure.io/fedora-atomic.git
+
+# Initialize a new empty repository in /srv/repo/ directory in archive mode
+RUN ostree --repo=repo init --mode=archive-z2
+
+# Expose default SimpleHTTPServer port and start SimpleHTTPServer
+EXPOSE 8000
+CMD python -m SimpleHTTPServer
+```
+
+> `Note:` If you are trying to compose OSTree on aarch64 or ppc64le then modify the first line of your Dockerfile from `FROM fedora:27` to `FROM docker.io/fedora:27`. This is needed because Docker first tries to pull in the Fedora 27 image from registry.fedoraproject.org which only contains x86_64 images for now.
+
+## Build Docker image and run container
+
+Now, we will build a Docker image from the Dockerfile we just created. In this example, our Dockerfile is available in the current directory.
+
+```
+$ sudo docker build --rm -t $USER/atomicrepo .
+```
+
+The Docker image has been created in the system which can be accessed with name $USER/atomicrepo. To view available images, run:
+
+```
+$ sudo docker images
+```
+
+Launch a container in the background from image $USER/atomicrepo we just created which listens on port 8000.
+
+```
+$ sudo docker run --privileged -d -p 8000:8000 --name atomicrepo $USER/atomicrepo
+```
+
+We now have a running container with SimpleHTTPServer running inside it. Let's get shell access to this container to work further:
+
+```
+$ sudo docker exec -it atomicrepo bash
+```
+
+We are now inside a running container with current working directory set to /srv.
+
+```
+# pwd
+/srv
+```
+
+## Composing OSTree repository
+
+While building the image, we cloned fedora-atomic repository into `/srv/` directory. We will use content from the fedora-atomic repository as reference to build custom OSTree. Let's checkout f27 branch of the fedora-atomic repo:
+
+```
+# cd fedora-atomic
+# git checkout f27
+```
+
+File `fedora-atomic-host-base.json` in the `fedora-atomic` repo is the [treefile](https://rpm-ostree.readthedocs.io/en/latest/manual/treefile/) containing details about the target system. It has a list of packages which will be composed into our new OSTree. Add or remove package names as required in your custom OSTree repo. In this example, I have added `elfutils` in the `packages` list.
```

Comment:
i'm pretty sure we want to tell people to use: `fedora-atomic-host.json` insted of `fedora-atomic-host-base.json`. The base/updates/testing json files were just put there as a temporary workaround.

Reply:
Among all json files mentioned in fedora-atomic repo, I see that only fedora-atomic-host-base.json contains details of what treefile should have. Rest of the json files mainly include fedora-atomic-host-base.json file. fedora-atomic-host-base.json file is self-sufficient for composing OSTree repo and hence I mentioned it to avoid any extra confusion to readers.

Also, value of "ref:" field is assigned in both fedora-atomic-host-base.json and fedora-atomic-host.json.This made me a bit confused that exactly which one will be used by rpm-ostree during compose process.

Reply:
yeah what I'm trying to do is have this article not reference a file that will go away in the near future. The master branch shows you what the files should look like: https://pagure.io/fedora-atomic/tree/master